python/dolfinx/fem/petsc.py

```diff
@@ -214,201 +339,97 @@ def _assemble_vector_vec(b: PETSc.Vec, L: Form, constants=None, coeffs=None) ->
-    else:
-        x0_local = []
-        x0_sub = [None] * len(maps)
-
-    constants_L = (
-        [form and _pack_constants(form._cpp_object) for form in L]
-        if constants_L is None
-        else constants_L
-    )
-
-    coeffs_L = (
-        [{} if form is None else _pack_coefficients(form._cpp_object) for form in L]
-        if coeffs_L is None
-        else coeffs_L
-    )
-
-    constants_a = (
-        [
-            [
-                _pack_constants(form._cpp_object)
-                if form is not None
-                else np.array([], dtype=PETSc.ScalarType)
-                for form in forms
-            ]
-            for forms in a
+def assemble_matrix(
+    a: typing.Union[Form, Iterable[Iterable[Form]]],
+    bcs: typing.Optional[Iterable[DirichletBC]] = None,
+    diag: float = 1,
+    constants: typing.Optional[
+        typing.Union[Iterable[np.ndarray], Iterable[Iterable[np.ndarray]]]
+    ] = None,
+    coeffs: typing.Optional[
+        typing.Union[
+            dict[tuple[IntegralType, int], npt.NDArray],
+            Iterable[dict[tuple[IntegralType, int], npt.NDArray]],
         ]
-        if constants_a is None
-        else constants_a
-    )
+    ] = None,
+    kind=None,
+):
+    """Assemble a bilinear form into a matrix.
 
-    coeffs_a = (
-        [
-            [{} if form is None else _pack_coefficients(form._cpp_object) for form in forms]
-            for forms in a
-        ]
-        if coeffs_a is None
-        else coeffs_a
-    )
+    The following cases are supported:
 
-    _bcs = [bc._cpp_object for bc in bcs]
-    bcs1 = _bcs_by_block(_extract_spaces(a, 1), _bcs)
-    b_local = _cpp.la.petsc.get_local_vectors(b, maps)
-    for b_sub, L_sub, a_sub, const_L, coeff_L, const_a, coeff_a in zip(
-        b_local, L, a, constants_L, coeffs_L, constants_a, coeffs_a
-    ):
-        _cpp.fem.assemble_vector(b_sub, L_sub._cpp_object, const_L, coeff_L)
-        _a_sub = [None if form is None else form._cpp_object for form in a_sub]
-        _cpp.fem.apply_lifting(b_sub, _a_sub, const_a, coeff_a, bcs1, x0_local, alpha)
-
-    _cpp.la.petsc.scatter_local_vectors(b, b_local, maps)
-    b.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
-
-    bcs0 = _bcs_by_block(_extract_spaces(L), _bcs)
-    offset = 0
-    b_array = b.getArray(readonly=False)
-    for submap, bcs, _x0 in zip(maps, bcs0, x0_sub):
-        size = submap[0].size_local * submap[1]
-        for bc in bcs:
-            bc.set(b_array[offset : offset + size], _x0, alpha)
-        offset += size
+    1. If ``a`` is a single bilinear form, the form is assembled
+       into PETSc matrix of type ``kind``.
+    #. If ``a`` is a rectangular array of forms the forms in ``a`` are
+       assembled into a matrix such that::
 
-    return b
+            A = [a_00 ... a_0n]
+                [a_10 ... a_1n]
+                [     ...     ]
+                [a_m0 ..  a_mn]
 
+       a. If ``kind`` is a ``PETSc.Mat.Type`` (other than
```

Comment:
Couldn't easily follow the logic in the brackets at the back end of this statement.